extensions/amp-a4a/0.1/test/test-signature-verifier.js

```diff
@@ -0,0 +1,479 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// TODO(@erwinmombay, #11081): eslint-disable-next-line needed due to linter bug
+
+import * as sinon from 'sinon';
+
+import {dev, user} from '../../../../src/log';
+import {base64EncodeFromBytes} from '../../../../src/utils/base64';
+import {utf8EncodeSync} from '../../../../src/utils/bytes';
+import {SignatureVerifier, VerificationStatus} from '../signature-verifier';
+
+const networkFailure = {throws: new TypeError('Failed to fetch')};
+const noop = () => {};
+
+describes.fakeWin('SignatureVerifier', {amp: true}, env => {
+
+  let mockDevError;
+  beforeEach(() => {
+    mockDevError = env.sandbox.mock(dev()).expects('error');
+    mockDevError.never();
+  });
+  afterEach(() => {
+    mockDevError.verify();
+  });
+
+  /** @param {string} signingServiceName */
+  const expectSigningServiceError = signingServiceName => {
+    mockDevError.withExactArgs('AMP-A4A', sinon.match(signingServiceName))
+        .once();
+  };
+
+  const creative1 = utf8EncodeSync('Hello world!');
+  const creative2 = utf8EncodeSync('This is a <em>test</em> creative.');
+
+  it('should make no network requests when crypto is unavailable', () => {
+    env.sandbox.stub(env.win, 'crypto', undefined);
+    env.expectFetch('*', {throws: new Error('no network requests allowed')});
+    const verifier = new SignatureVerifier(env.win);
+    verifier.loadKeyset('service-1');
+    return verifier
+        .verifyCreativeAndSignature(
+            // eslint-disable-next-line indent
+            'service-1', 'key-1', creative1, new Uint8Array(256), noop)
+        .then(status => {
+          expect(status).to.equal(VerificationStatus.UNVERIFIED);
+          expect(env.fetchMock.called()).to.be.false;
+        });
+  });
+
+  const crypto = window.crypto || window.msCrypto;
+  if (!crypto) {
+    return;
+  }
+  const subtle = crypto.subtle || crypto.webkitSubtle;
+  if (!subtle) {
+    return;
+  }
+
+  describe('when crypto is available', () => {
```

Comment:
That `verifyCreativeAndSignature` calls itself is really an implementation detail to make the code more concise, so I think a test that spied on it directly would be overly fragile. It seems like what we really want to assert is that the verifier will not make more than two HTTP requests for a single kid; for that, see `should not make more network requests retrying a nonexistent kid`.